Subscribe refactor

erikkemperman opened the pull request:
Ref:
https://github.com/ReactiveX/RxPY/issues/430
https://github.com/ReactiveX/RxPY/issues/293#issuecomment-457502444

I'm a bit conflicted about submitting this -- on the one hand, it seems like too big a patch just before release, on the other hand it would change the API surface and therefore this is probably also something that could not land just after a major release...

But personally I think something like this would improve the API, and if I understand the earlier discussion properly this might be pretty close to consensus to how things ought to be. It seems to me the main reason this wasn't attempted earlier is that it would be quite a tedious thing to do. And so it was :-)

But I got things working, as far as I'm aware, and thus I've decided to just throw this out there, curious what people think.

The idea is basically to adapt the principal user-facing API (`Observable`) to promote the "callback style" of subscribing to an observable. The "observer style" is still supported, and used internally in two ways, but only as implementation details (namely, the operators and the observable sub-class hierarchy).

In other words, the observer style of subscribing only remains in places that could be adapted without touching the observable API.

For what it's worth, the changes here plenty, but they are almost entirely "mechanical":

- Get rid of polymorphism in `Observable.subscribe`, change the signature to only accept an observer. All invokations of `subscribe` that were using callbacks change to `subscribe_`.

- Rename `subscribe` to `subscribe_observer` and rename `subscribe_` to `subscribe`.

- The observer argument to `subscribe_observer` does not have to be optional, if it is not given we may as well invoke the callback-style variant with no arguments. So, considering it compulsory, we can turn things around by moving `Observable.subscribe_observer(observer)` into `Observer.subscribe_to(observable)`.

The last step means that the typing / ABC for Observable now only knows about the callback style. To use the observer style, you have to first have an instance of Observer and subscribe via that instance.

I've run the whole thing through Travis to make sure each of these steps resulted in a passing build, as evidenced by the commit history below.

dbrattli commented:
Nice work! The thing I'm having problem digesting is the flipped `subscribe_to` that breaks with Rx especially when it needs to be in the abc. Would it perhaps be possible to have it as a plain old function instead (since they all look the same)? Thus we could try eventually get rid of the observers internally and only have callback (but not having to taint the observer interface in the process). E.g having a `subscribe_observer` utility ...

erikkemperman commented:
Hmm, yes, I guess a standalone utility might work — although simply removing it from the public observer might be sufficient? I think none of this relies on it being in typing.Observer.

I’ll take a look at it!

erikkemperman commented:
FYI, I have removed the `subscribe_to` from typing and ABC -- so the internal Observer classes still have it, but it's not a requirement for custom implementations.

erikkemperman commented:
Got rid of a bunch of typing errors. Was hoping that I wouldn't have to revisit all these files again, but can't be helped in this case. Removed `subscribe_to` from core.Observable and added it as static function `internal.util.subscribe(observable, observer)`.

dbrattli commented:
This is beginning to look really nice! I think it should be possible to also let `_subscribe_core` take callbacks as well? I.e `self._subscribe_core(auto_detach_obs.on_next, auto_detach_obs.on_error, auto_detach_obs.on_completed, , scheduler)`.

erikkemperman commented:
Yes, that was the intention. But I think maybe not in this PR? It’s already rather huge! My goal here was to get the API-touching part in, the internal bits could be done later and should not need breaking changes... I mean, I’d be happy to try it but I’m worried it would make this even more annoying to review.

dbrattli commented:
I was thinking about just in `Observable.py`, but I see that we then need to wrap the callbacks back into an observer to avoid the next huge rewrite (and that is where I stopped last time and gave up). So you are right, let's wait with that step.

dbrattli commented on notebooks/reactivex.io/startup.py:
```diff
@@ -88,7 +88,7 @@ def on_completed(self)   : return self._on('cmpl', 'fin')
     # required for e.g. .multicast:
     obs = Subscriber(src, **kw)
-    subscription = src.subscribe(obs)
+    subscription = obs.subscribe_to(src)
```
Revert?

erikkemperman replied:
Ah, yes, that's not right. Thanks!

And actually Subscriber isn't strictly speaking a subclass of `Observer` here, so I changed it to just pass the callbacks from `obs` to `src.subscribe`.

My impressions is the notebooks need to still be aligned to v3, anyway, is that correct?

I've amended, and also found another quirk in delaywithmapper (I think it was broken before, apparently not covered by tests).

erikkemperman commented on rx/core/observable/observable.py:
```diff
@@ -23,88 +23,57 @@ class Observable(typing.Observable):
+                                   observer.on_completed,
+                                   scheduler)
+
+        if callable(self._subscribe_observer):
```
Ah, yes indeed. Do you think it would be better to check if not None and then just try the call, so an exception will be raised if not callable?

erikkemperman replied:
I think you're right, and I've changed it -- now it follows the previous pattern of just checking if the function is set and then attempt to call it. Thanks!!

erikkemperman commented:
@dbrattli I can see why you gave up on the internal part... I mostly got it to work, but it was pretty annoying. Not really difficult, but *a lot*!

https://github.com/erikkemperman/RxPY/commit/3456a1940802bfaff34e244019f441a6cf6a96eb

I think it's probably better to keep it on the shelf for the moment, though. However, while doing this I ran into a few things that looked off to me and might be legitimate bugs... I'm going to have to leave it here for today, as soon as I have some time I'll discuss those separately.

erikkemperman commented:
(Rebased onto latest master to resolve merge conflict)

dbrattli commented on rx/core/observable/observable.py:
```diff
@@ -116,11 +85,11 @@ def subscribe_(self,
         """
 
-        auto_detach_observer = AutoDetachObserver(on_next, on_error, on_completed)
+        auto_detach_obs = AutoDetachObserver(on_next, on_error, on_completed)
```
Observers are abbreviated as obv. Observables are abbreviated as obs.

erikkemperman commented:
I rebased this on latest master to resolve merge conflicts. Had to do it twice, did not notice the b5 release. Sorry about all the force-pushes. I also changed the observer abbreviation to `obv` as you mentioned.

For some reason, mypy is raising new warnings now, that were not there on this branch before master, and neither were they there on the master branch. I don't think it's anything particularly scary, but I hope to have some time for a better look at it.

kormang commented:
Hi!
I also see one more problem with subscribe, PTAL at #565 .

dbrattli commented:
I'm closing this PR since it's to large and has drifted too far from the current master. I also think the current master fixes some of the same issues ref mypy typing.
